Blockly.Xml.domToWorkspaceAlign(): loads blocks aligned to existing blocks

qnoirhomme opened the pull request:
Updated Blockly.Xml.domToWorkspace to load previously saved "blocks"
below the existing blocks

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly/822)
<!-- Reviewable:end -->

qnoirhomme commented:
I just signed the CLA!

AnmAtAnm commented:
Since this is a breaking behavior change, this should be moved into a separate function.  Alternatively, it could be default disabled with an optional parameter, but I'm not as fond of that solution.

Further, the code makes assumptions about the location of the blocks that are about to be loaded. It is quite possible the loaded blocks are not just below y=0 or have a different x alignment relative to the existing blocks.  For this to work, I believe you would have to load the blocks into the workspace (without the offset), keeping a list of these new blocks, and then shift their position once you know the resulting positions and sizes.

If domToWorkspace() returned the list of newly added blocks, it would be possible to implement this behavior in a wrapper function.

RoboErikG commented on core/xml.js:
```diff
@@ -301,6 +301,25 @@ Blockly.Xml.domToWorkspace = function(xml, workspace) {
+  var bottomY =0; // container for the y of the bottom block
+  // sort the blocks in reverse order
+  if (blocks.length > 1) {
+	 blocks.sort(function(a, b) { // sort the block acording to their position and height
```
Since you're only interested in the largest y value sorting is unnecessary. It would be faster to just iterate through the list and store the largest y + height value.

qnoirhomme replied:
Indeed, thanks.
If I want to align the block, as suggested by AnmAtAnm, I will also need the x position, so a loop through the blocks will be better.

qnoirhomme replied:
Doing the change, I was wondering if, in RTL mode, the block position was only mirrored before to be added or if the block position was already reversed in the xml file. I assumed the first option, based on the existing code. This has importance for the getRelativeToSurfaceXY which won't give me the 'x' I look for otherwise (I would have to look for x + width)

RoboErikG commented on core/xml.js:
```diff
@@ -311,7 +330,7 @@ Blockly.Xml.domToWorkspace = function(xml, workspace) {
       var block = Blockly.Xml.domToBlock(xmlChild, workspace);
       var blockX = parseInt(xmlChild.getAttribute('x'), 10);
-      var blockY = parseInt(xmlChild.getAttribute('y'), 10);
+      var blockY = parseInt(xmlChild.getAttribute('y'), 10) + bottomY; // add bottomY to be below all blocks
```
As Andrew pointed out, some blocks may have a negative 'y' position or the new blocks might start at a very high y position. You'll want your new method to keep a list of all the blocks and the lowest Y value and then add the difference between the lowest y here and the highest y above to every block.

You also probably want to adjust the x values so the new blocks are left aligned with the existing blocks or you could have the same issue of the blocks being very far apart horizontally.

qnoirhomme replied:
Indeed, thanks. 
I now keep track of the block and their position before the move, compute the offset and move the blocks. That way I move the block only once.

qnoirhomme commented:
Thanks for the comment AnmAtAnm!
I think, for a wrapper function I would also need the list of the previously existing blocks, otherwise I may compute the position of the existing block. Instead I thought about creating a new function domToWorkspaceAlign.
I guess I should also update the unit testing then.

AnmAtAnm commented on core/xml.js:
```diff
@@ -311,7 +312,7 @@ Blockly.Xml.domToWorkspace = function(xml, workspace) {
       var block = Blockly.Xml.domToBlock(xmlChild, workspace);
       var blockX = parseInt(xmlChild.getAttribute('x'), 10);
-      var blockY = parseInt(xmlChild.getAttribute('y'), 10);
+      var blockY = parseInt(xmlChild.getAttribute('y'), 10); // add bottomY to be below all blocks
```
Remove comment.

AnmAtAnm commented on core/xml.js:
```diff
@@ -332,6 +333,122 @@ Blockly.Xml.domToWorkspace = function(xml, workspace) {
+ * @param {!Element} xml XML DOM.
+ * @param {!Blockly.Workspace} workspace The workspace.
+ */
+Blockly.Xml.domToWorkspaceAlign = function(xml, workspace) {
```
I prefer name appendDomToWorkspace(..) for this new function.

qnoirhomme commented:
Instead of factoring domToWorkspace into smaller functions, I slightly modified the function to accept an "offset" parameter. The appendDomToWorkspace now computes the offset and then call domToWorkspace. No more double coding there. domToWorkspace can still perfectly work without including the extra parameter.
I had some trouble with unit testing as they run in headless mode and therefore there is no height value. I added an extra check and now it is running fine.

AnmAtAnm commented on tests/jsunit/xml_test.js:
```diff
@@ -75,8 +77,35 @@ function test_domToWorkspace() {
+        '  <block type="test_block" inline="true" x="21" y="23">' +
+        '  </block>' +
+        '</xml>');
+    var workspaceA = new Blockly.Workspace();
```
This is a headless workspace.  Make sure you include a rendered workspace test, also.

qnoirhomme replied:
I tried to use a WorkspaceSvg, instead of a Workspace for that commit but then I had trouble with getCanvas. I could not fine an example of the use of WorkspaceSvg in unit testing. Could you point one for me? 
Thank you for your help!

AnmAtAnm replied:
Unfortunately, our test coverage of this web version of Blockly is pretty sparse, especially for things that are rendered, like any of the SVG classes.

WorkspaceSVG @extends Blockly.Workspace, and Blockly.mainWorkspace is almost always a WorkspaceSVG object.  There are several examples of WorkspaceSVG.getCanvas() and .getBubbleCanvas() in the WorkspaceSVG class itself.  What problem were you having?

qnoirhomme replied:
I try to directly create a workspaceSvg
var workspaceB = new Blockly.WorkspaceSvg({toolbox: XML_TOOLBOX});
where var XML_TOOLBOX = ['<xml id="toolbox" style="display: none">',
 '  <category name="Logic" colour="210">',
 '     <block type="controls_if"></block>',
 '     <block type="logic_compare"></block>',
 '     <block type="logic_operation"></block>',
 '     <block type="logic_negate"></block>',
 '     <block type="logic_boolean"></block>',
 '     <block type="logic_null"></block>',
 '     <block type="logic_ternary"></block>',
 '   </category>',
 '</xml>'].join('\n');  

Then I inject a block from an xml snippet
Blockly.Xml.domToWorkspace(dom, workspaceB);
But I get 
this.workspace.getCanvas(...) is undefined
> Blockly.BlockSvg.prototype.initSvg at file:///blockly/core/block_svg.js:147:5
> Blockly.Xml.domToBlock at file:///blockly/core/xml.js:435:9
> Blockly.Xml.domToWorkspace

I guess the way I call the workspaceSvg is too simple but I tried to avoid the html version of it Blockly.Inject().

qnoirhomme replied:
I cannot see anything in the toolbox in my previous message but this is just a copy of a the xml tags with the logic category inside (and the related blocks)

AnmAtAnm replied:
I'm not clear why you need to create a Workspace object, when a workspace parameter is passed into both methods.

AnmAtAnm replied:
Oh...  I understand, now.
Add your rendered tests to [tests/workspace_svg/workspace_svg_test.js](https://github.com/google/blockly/tree/develop/tests/workspace_svg/workspace_svg_test.js).

Apologies for the confusion.

qnoirhomme replied:
Thank you!

AnmAtAnm commented on tests/jsunit/xml_test.js:
```diff
@@ -75,8 +77,35 @@ function test_domToWorkspace() {
         '</xml>');
     var workspace = new Blockly.Workspace();
-    Blockly.Xml.domToWorkspace(dom, workspace);
+	Blockly.Xml.domToWorkspace(dom, workspace);
```
Fix the incorrect indents.

AnmAtAnm commented on core/xml.js:
```diff
@@ -274,15 +274,21 @@ Blockly.Xml.textToDom = function(text) {
   }
+  if (offset === undefined) { offset = [0,0];}
+  if (offset.constructor !== Array) { offset = [0,0];}
+  if (offset.length !== 2) { offset = [0,0];}
```
These last two cases are developer errors. Throw an exception.

qnoirhomme commented on core/xml.js:
```diff
@@ -274,15 +274,21 @@ Blockly.Xml.textToDom = function(text) {
+ * @param {Array} offset array with x and y offset for the new blocks, default [0,0]
  */
-Blockly.Xml.domToWorkspace = function(xml, workspace) {
+Blockly.Xml.domToWorkspace = function(xml, workspace, offset) {
```
Thank you for all your comments!
I will adapt my code before the next commit. I already switched to develop branch.

qnoirhomme replied:
I do know where the incoming blocks are located that's the second loop in the function which parse the xml file for the block locations.

AnmAtAnm replied:
Ahhh.. Cool.

qnoirhomme commented on core/xml.js:
```diff
@@ -332,6 +338,71 @@ Blockly.Xml.domToWorkspace = function(xml, workspace) {
+  if (blocksCount > 0) {
+	// initialize the position of the corners using position 
+	// of the first block
+	var bHW = (("height" in blocks[0])? blocks[0].getHeightWidth() : {height: 0, width: 0});
```
I am wondering what is the best solution. getBlocksBoudingBox add a Tab in case of open connection which slightly misaligned the code. As a workaround, I set Blockly.BlockSvg.TAB_WIDTH to 0 before to call the function and then back to its normal value. This help me to save a few lines of code. I hope it is ok with you.

AnmAtAnm replied:
@picklesrus Is the TAB width in a unit appropriate for this case, with respect to zoom? Should the getBlocksBoundingBox() be updated to include this width? It looks like it should for zoomToFit(), but we also use it in getTopLevelWorkspaceMetrics_().

picklesrus replied:
I'm not sure I fully understand the questions... so I'll just try to give as much info as I can to hopefully answer some of them :)

All of the height/width info that is cached in blocks (and therefore the result of the bounding box calculation) is always done at scale of 1.  The zoom is done by the scale transformation on the svg group and we don't update our internal references to sizes to reflect that.

to calculate content width and height getTopLevelWorkspaceMetrics applies scale after it gets the bounding box:https://github.com/google/blockly/blob/f5f6d69d5ef86d92e249054004c39690b0961f61/core/workspace_svg.js#L1545

As for tab, getBlocksBoundingBox should already include the tab because it calls into block's getBoundingRectangle which adds a tab: https://github.com/google/blockly/blob/f5f6d69d5ef86d92e249054004c39690b0961f61/core/block_svg.js#L491

AnmAtAnm replied:
I think the issue is alignment. I see XXX case (assuming LTR):

 * The left-most existing block and the left-most new block lack an output connector. Then, all blocks align to that left edge.
 * The left-most existing block and the left-most new block both have an output connector (and no other top block is within TAB_WIDTH of that left edge). Then, all blocks also align to the same left edge.
 * The left-most existing block has a output connector, and new blocks do not.  Then, shift new blocks right by TAB_WIDTH.
 * The left-most existing block does not have an output connector, but the new blocks do. Then, shift the new blocks left by TAB_WIDTH.
 * The left-most existing block has a tab, but an also another existing block without a tab is within TAB_WIDTH. Then.... ?
 * The left-most new block has a tab, but the also another new block without a tab is within TAB_WIDTH. Then...?

I would accept a pull request that doesn't handle the last two case.

qnoirhomme replied:
I suppose you assume that all blocks have been loaded in the workspace so we can use getBlocksBoundingBox() and getBoundingRectangle().
In the current situation, the new blocks are not affected by the TAB since this is not included in the xml position. From what I understand from @picklesrus, the way I suggest:
var savetab = Blockly.BlockSvg.TAB_WIDTH;
Blockly.BlockSvg.TAB_WIDTH = 0;
var bbox = workspace.getBlocksBoundingBox();
Blockly.BlockSvg.TAB_WIDTH = savetab;  

may only be problematic if the user is doing something else requesting TAB_WIDTH at the time he is loading a file  with the domToWorkspace function. Doing this will take care of case 5.

AnmAtAnm replied:
JavaScript is a single threaded, so is mostly safe.  Do this in a try finally block, to protect against exceptions:

```
var savetab = Blockly.BlockSvg.TAB_WIDTH;
try {
  Blockly.BlockSvg.TAB_WIDTH = 0;
  var bbox = workspace.getBlocksBoundingBox();
} finally {
  Blockly.BlockSvg.TAB_WIDTH = savetab;
}
```

qnoirhomme replied:
I am just wondering the use of the try finally block. I need bbox later in my code but if an error occurs, bbox will not be defined. I should run most of the code then in the try block or use a try catch finally block?

qnoirhomme commented:
Thanks for all your comments. I modified appendDomToWorkspace to read an array of new block ids from domToWorkspace and then move these new blocks if needed. 
I added a test function in workspace_svg_test.

AnmAtAnm commented on core/xml.js:
```diff
@@ -334,9 +338,68 @@ Blockly.Xml.domToWorkspace = function(xml, workspace) {
+Blockly.Xml.appendDomToWorkspace = function(xml, workspace) {
+  // first check if we have a workspaceSvg otherwise the block have no shape
+  // and the position does not matter
+  //if (typeof workspace.getBlocksBoundingBox() === undefined) {
```
Remove unused line.

AnmAtAnm commented:
It looks like you have abad merge that is editting several other files: accessible/libs/README, accessible/libs/angular2-polyfills.min.js, av12e0ds.5h5.txt

AnmAtAnm commented on core/xml.js:
```diff
@@ -334,9 +338,68 @@ Blockly.Xml.domToWorkspace = function(xml, workspace) {
 
 /**
+ * Decode an XML DOM and create blocks on the workspace.
+ * Append the new blocks at the bottom of the existing blocks
```
"Positions the new blocks immediately below any existing blocks, such that the left-most blocks are aligned."

AnmAtAnm commented on tests/jsunit/xml_test.js:
```diff
@@ -77,9 +78,35 @@ function test_domToWorkspace() {
+    Blockly.Xml.appendDomToWorkspace(dom, workspace);
+    assertEquals('Block count', 1, workspace.getAllBlocks().length);
+    Blockly.Xml.appendDomToWorkspace(dom, workspace);
+    assertEquals('Block count', 2, workspace.getAllBlocks().length);
```
Test the position of this new block.  Is it where you expect it to be?

AnmAtAnm commented on core/xml.js:
```diff
@@ -279,6 +279,7 @@ Blockly.Xml.textToDom = function(text) {
  * Decode an XML DOM and create blocks on the workspace.
  * @param {!Element} xml XML DOM.
  * @param {!Blockly.Workspace} workspace The workspace.
+ * @return {Array.<string>} An array containing new block ids.
```
Match the return value of domToWorkspace:
```
* @return {Array.<string>} An array containing new block ids.
```

qnoirhomme replied:
Indeed, thank you.

AnmAtAnm commented on core/xml.js:
```diff
@@ -334,9 +338,68 @@ Blockly.Xml.domToWorkspace = function(xml, workspace) {
+      var farY = bbox.y + bbox.height;
+      var topX = bbox.x;
+      // load the new blocks into the workspace and get the ids of the new blocks
+      var newblockids = Blockly.Xml.domToWorkspace(xml,workspace);
```
You have three calls to `domToWorkspace()`.   I think these can be reordered into one.

```
var blocks = workspace.getTopBlocks(true);  // Maybe rename priorBlocks
var bbox;
if (workspace.hasOwnProperty('scale')) {
  var savetab = Blockly.BlockSvg.TAB_WIDTH;
  try {
    Blockly.BlockSvg.TAB_WIDTH = 0;
    bbox = workspace.getBlocksBoundingBox();
  } finally {
    Blockly.BlockSvg.TAB_WIDTH = savetab;
  }
}
var newBlockIds = Blockly.Xml.domToWorkspace(xml,workspace);  // Camel case var name
if (bbox && bbox.height) {
  var offsetY = 0; // offset to add to y of the new block
  . . .
}
// No need for else statements.
return newBlockIds;  // Adding return to match domtoWorkspace()
```

qnoirhomme replied:
Nice, thank you!

AnmAtAnm commented on accessible/libs/README:
```diff
@@ -1,15 +1,15 @@
-Required by Angular2, for Javascript files.
-Fetched from https://github.com/paulmillr/es6-shim
-The current version is 0.35.1.
+This folder contains the following dependencies for accessible Blockly:
```
Revert file.

AnmAtAnm commented on accessible/libs/angular2-polyfills.min.js:
```diff
@@ -1,24 +1,24 @@
-OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
-THE SOFTWARE.
-**/
+/**
```
Revert file.

AnmAtAnm commented on av12e0ds.5h5.txt:
```diff
@@ -0,0 +1,196 @@
+﻿Merge remote-tracking branch 'refs/remotes/google/develop' into Change_domToWorkspace_QN
```
Revert file.

AnmAtAnm commented on msg/json/ja.json:
```diff
@@ -7,7 +7,8 @@
 			"Otokoume",
```
Revert file.

AnmAtAnm commented on msg/json/lv.json:
```diff
@@ -0,0 +1,313 @@
+{
```
Revert file.

qnoirhomme commented:
Thank you again for all the comments. I copy and paste all develop before to add my changes so, there are changes from my previous version but hopefully not from the current develop branch (besides the expected ones).

AnmAtAnm commented:
I think the easiest way forward right now is to start a new branch and new pull request.  Start clean from the current `develop` head, then copy and paste the two functions and the test (and anything else that is needed by them).

The alternatives are an ugly rebase to remove all the unrelated commits, or manually copying all the unrelated files. In the latter case, I'd be concerned we'd accidentally carry over master commits in the merge history.

I could do it myself, but I think it is best to make sure you get the credit for it.  Reference this PR (#822) when you re-request a pull with the new branch.

I'm looking forward to getting this in.  I know a couple of places where we can use it.  Thanks for all the effort.
